MCR-2891 Object History Permission

Hammer1279 opened the pull request:
- add history view check to listRevisions
- add history read check to requestVersionedObject

[Link to jira](https://mycore.atlassian.net/browse/MCR-2891).

Hammer1279 committed:
MCR-2891 Object History Permission
- add history view check to listRevisions
- add history read check to requestVersionedObject

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

Hammer1279 commented on mycore-base/src/main/java/org/mycore/datamodel/common/MCRXMLMetadataManager.java:
```diff
@@ -164,7 +166,8 @@ public MCRContent retrieveContent(MCRObjectID mcrid, String revision) throws IOE
     public List<? extends MCRAbstractMetadataVersion<?>> listRevisions(MCRObjectID id) throws IOException {
-        return IMPLEMENTATION.listRevisions(id);
+        return MCRAccessManager.checkPermission(id, MCRAccessManager.PERMISSION_HISTORY_VIEW)
+            ? IMPLEMENTATION.listRevisions(id) : Collections.emptyList();
```
But this is a read only method?

rsteph-de replied:
Do not check Permissions in MCRXMLMetadataManager
Permission check should be implemented where the method is called (REST-API, URI-Resolver, MCRObjectServlet, ???)

rsteph-de commented on mycore-base/src/main/java/org/mycore/frontend/servlets/MCRObjectServlet.java:
```diff
@@ -100,6 +101,14 @@ private MCRContent requestLocalObject(MCRObjectID mcrid, final HttpServletRespon
+            final MCRSession currentSession = MCRSessionMgr.getCurrentSession();
+            resp.sendError(
+                HttpServletResponse.SC_UNAUTHORIZED,
+                getErrorI18N(I18N_ERROR_PREFIX, "accessDenied", mcrid.toString() + " " + rev,
```
use a better name for "accessDenied", something like "accessToVersionDenied"
and provide a proper error message

rsteph-de reviewed (changes requested):
Sebastian and I looked into the code, I will add more implementation notes to the JIRA-ticket

Hammer1279 committed:
MCR-2891 Object History Permission
- permission check in URI Resolver instead of XML MetadataManager
- use own error message for versioned lack of permissions

Hammer1279 commented:
~~I think I'll have to recreate the branch... this looks very wrong...~~
Resolved, but leaving this here to explain what happened.

Hammer1279 committed:
MCR-2891 Object History Permission
- add required permission to rest api

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

Hammer1279 commented:
Note for the missing i18n for anything except for german and english:
While I am able to edit the i18n, it would ruin the entire formatting of all those files, so I had to edit them by hand instead of with tools.
That raises the question if it is important to have them formatted to be easily human readable if we mostly use tools anyways to edit them or if it is okay to have them all bunched together.

rsteph-de committed:
MCR-2891 move ACLcheck from MCRVersionInfoResolver to versions.xsl

rsteph-de committed:
MCR-2891 add empty new line in versions.xsl

rsteph-de committed:
MCR-2891 fix xsl typo

rsteph-de reviewed (approved):
looks good now after final changes  (acl:checkPermission in versions.xml)
